pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala

```diff
@@ -0,0 +1,62 @@
+package uk.ac.wellcome.platform.merger.rules
+import uk.ac.wellcome.models.work.internal.{
+  TransformedBaseWork,
+  UnidentifiedWork
+}
+import uk.ac.wellcome.platform.merger.rules.WorkFilters.WorkFilter
+
+/*
+ * MergeResult is the return type of a FieldMergeRule's `merge` method
+ * and contains both the new (merged) value for the field but also a list
+ * of sources which the rule would like the merger to redirect. It is up to
+ * the merger how to handle these.
+ */
+case class MergeResult[T](fieldData: T, redirects: Seq[TransformedBaseWork])
+
+/*
+ * A trait to extend in order to merge fields of the type member `Field`
+ *
+ * The implementor must provide the method `merge`, which takes a target
+ * work and a list of source works, and returns the new value of the field
+ * as well as a list of source works that this rule would like to be redirected.
+ *
+ * Many merge rules will want to apply different logic given conditions on
+ * both the target and the source works.
+ * The PartialRule trait is provided to achieve this: in addition to
+ * a `rule` that has the same signature as the `merge` method, the implementor
+ * must provide predicates for valid targets and sources.
+ *
+ * Because PartialRule is a PartialFunction, the `rule` will never be called
+ * for targets/sources that don't satisfy these predicates, and we gain things
+ * like `orElse` and `andThen` for free.
+ */
+trait FieldMergeRule {
+  protected final type Params = (UnidentifiedWork, Seq[TransformedBaseWork])
+  protected type FieldData
+
+  def merge(target: UnidentifiedWork,
+            sources: Seq[TransformedBaseWork]): MergeResult[FieldData]
+
+  protected val identityOnTarget: PartialFunction[Params, UnidentifiedWork] = {
+    case (target, _) => target
+  }
+
+  protected trait PartialRule extends PartialFunction[Params, FieldData] {
+    val isDefinedForTarget: WorkFilter
+    val isDefinedForSource: WorkFilter
```

Comment:
A bit unsatisfied with the naming of these vars :thinking: 

It's not whether the rule is "defined", more whether it "applies", or something. `appliesToTarget` / `appliesToSource`?

Reply:
Maybe also change `WorkFilter` to `WorkPredicate`? Or is that too much of a mouthful?

Reply:
(not really a filter as such, and already have `WorkFilter` type elsewhere which doesn't help)

Reply:
Hmmm, the naming is very deliberately a reference to the naming used by (and the general concept of) [`PartialFunction`](https://www.scala-lang.org/files/archive/api/2.12.10/scala/PartialFunction.html), as `PartialRule`s are no more than functions over _a subset of_ the product of the set of targets and the set of lists of sources. Especially given that I use `isDefinedAt` outside of the implementation I would prefer to keep the naming of these the same so as not to obfuscate the connection.

Completely agree for `WorkPredicate` though!